### PR TITLE
Remove writable JSON DOM

### PR DESCRIPTION
**Pull request by layomia**
Per conversation in https://github.com/dotnet/runtime/issues/30436, this PR removes the writable JSON dom.

Shipping a writable dom is on the roadmap for JSON, but is not prioritized at the moment. https://github.com/dotnet/runtime/issues/30436#issuecomment-535301281 outlines the next steps, which includes responding to UX study feedback and further design + API review. Iteration on this feature will probably be in a feature branch.

cc @ericstj, @joperezr, @ahsonkhan, @stephentoub

### CODE REVIEW

**Review by bartonjs — APPROVED** [review]

**Review by stephentoub — APPROVED** [review]
Thank you.

File: src/libraries/System.Text.Json/src/System/Text/Json/Document/JsonElement.cs [hunk 1]
```diff
@@ -414,26 +331,15 @@ public bool TryGetProperty(ReadOnlySpan<byte> utf8PropertyName, out JsonElement
...
-            if (_parent is JsonDocument document)
-            {
-                JsonTokenType type = TokenType;
-
-                return
-                    type == JsonTokenType.True ? true :
-                    type == JsonTokenType.False ? false :
-                    throw ThrowHelper.GetJsonElementWrongTypeException(nameof(Boolean), type);
-            }
+            // CheckValidInstance is redundant.  Asking for the type will
```
Comment (ahsonkhan) [inline src/libraries/System.Text.Json/src/System/Text/Json/Document/JsonElement.cs]: @bartonjs, should we still keep this to validate the `JsonElement elem = default` cases?

File: src/libraries/System.Text.Json/tests/JsonNumberTestData.cs [hunk 2]
```diff
@@ -2,11 +2,8 @@
 // The .NET Foundation licenses this file to you under the MIT license.
 // See the LICENSE file in the project root for more information.
 
-using System.Buffers;
 using System.Collections.Generic;
 using System.Globalization;
-using System.IO;
-using Newtonsoft.Json;
```
Comment (ahsonkhan) [inline src/libraries/System.Text.Json/tests/JsonNumberTestData.cs]: nit: Remove unrelated cleanup like this (especially in a file that would otherwise not be touched by your PR).

File: src/libraries/System.Text.Json/src/System/Text/Json/Document/JsonElement.cs [hunk 3]
```diff
@@ -54,21 +47,7 @@ private JsonTokenType TokenType
         /// <exception cref="ObjectDisposedException">
         ///   The parent <see cref="JsonDocument"/> has been disposed.
         /// </exception>
-        public JsonValueKind ValueKind
-        {
-            get
-            {
-                if (IsImmutable)
```
Comment (ahsonkhan) [inline src/libraries/System.Text.Json/src/System/Text/Json/Document/JsonElement.cs]: Remove the `IsImmutable` property as well (and from the ref)? It will never be true without `JsonNode` or a writable DOM equivalent.

**Review by ahsonkhan — APPROVED** [review]
Missed one API. Otherwise, looks good.

File: src/libraries/System.Text.Json/src/System/Text/Json/Document/JsonElement.cs (hunk 1 above)
Comment (bartonjs) [reply src/libraries/System.Text.Json/src/System/Text/Json/Document/JsonElement.cs]: I was going to make the same comment, then checked and saw that this is just restoring it exactly to what shipped in 3.1, so it's good.

**Review by joperezr — APPROVED** [review]
Thanks @layomia
